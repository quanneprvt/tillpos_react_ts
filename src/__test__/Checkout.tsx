import React from 'react';
import Checkout from "../modules/Checkout";
import { shallow, mount } from "enzyme";

test('renders learn react link', () => {
  let wrapper;
});
