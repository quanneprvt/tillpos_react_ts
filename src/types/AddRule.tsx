interface IRuleData {
  itemType: {
    vendor: string,
    quantity: string,
    amount: number,
    product: string
  },
  discountType: {
    amount: number,
    price: number
  }
}

export type { IRuleData };