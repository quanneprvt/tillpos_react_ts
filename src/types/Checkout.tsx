interface IDiscountResult {
  totalAmount: number;
  price: number;
  percent: number;
}

export type { IDiscountResult }