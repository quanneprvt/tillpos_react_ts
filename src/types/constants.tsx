import { ISelect } from "./Select";

interface IProductData {
  name: string;
  description: string;
  price: number;
}

interface IVendorData extends ISelect {}

interface ISelectCondition {
  GREATER_THAN: ISelect;
  GREATER_OR_EQUAL_THAN: ISelect;
  EQUAL: ISelect;
  LESS_THAN: ISelect;
  LESS_OR_EQUAL_THAN: ISelect;
}

interface IProductDiscountType {
  EACH: ISelect;
  ALL: ISelect;
}

interface IRuleMode {
  ADD: ISelect;
  EDIT: ISelect;
}

export type { IProductData, IVendorData, ISelectCondition, IProductDiscountType, IRuleMode };
