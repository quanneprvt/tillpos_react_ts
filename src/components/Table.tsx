import React from "react";
import {
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Paper
} from "@mui/material";
import { ITableProps } from "../types/table";

class TableComponent extends React.Component<ITableProps> {
  headers: Array<string>;

  constructor(props: ITableProps) {
    super(props);
    this.headers = this.getDistinctField();
  }

  getDistinctField(): Array<string> {
    let arr: Array<string> = [];
    if (this.props.data) {
      for (let i = 0; i < this.props.data.length; i++) {
        arr.push(...Object.keys(this.props.data[i]));
      }
    }
    return [...Array.from(new Set(arr))];
  }

  render() {
    return (
      <TableContainer component={Paper}>
        <Table sx={{ minWidth: 200 }} aria-label="simple table">
          <TableHead>
            <TableRow>
              {this.headers.map((header, index) => (
                <TableCell
                  align="center"
                  key={index}
                  sx={{ "&": { textTransform: "capitalize" } }}
                >
                  {header}
                </TableCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {this.props.data &&
              this.props.data.map((row, rowIndex) => (
                <TableRow
                  key={rowIndex}
                  sx={{ "&:last-child td, &:last-child th": { border: 0 } }}
                >
                  {this.headers.map((header, valueIndex) => (
                    <TableCell align="center" key={valueIndex}>
                      {row[header]}
                    </TableCell>
                  ))}
                </TableRow>
              ))}
          </TableBody>
        </Table>
      </TableContainer>
    );
  }
}

export default TableComponent;
