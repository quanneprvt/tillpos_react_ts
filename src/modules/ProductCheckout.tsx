import React from "react";
//component
import Table from "../components/Table";
import Select from "../components/Select";
import Dialog from "../components/Dialog";
import List from "../components/List";
import ListItemCollapse from "../components/ListItemCollapse";
import { Button, Container, ListItemText } from "@mui/material";
import { IItemAdd } from "../types/Table";
//module
import AddRule from "./AddRule";
//constant
import { tableData, vendorData } from "../constants/index";
interface IItemProduct extends IItemAdd {
  count: number;
}

interface IItemsCheckout {
  vendor: string;
  items: Array<IItemProduct>;
}

interface IProps {}

interface IStates {
  addedItems: Array<IItemsCheckout>;
}

class ProductCheckoutModule extends React.Component<IProps, IStates> {
  selectedVendor: string | undefined;
  constructor(props: IProps) {
    super(props);
    this.selectedVendor = vendorData[0].value;
    this.state = {
      addedItems: []
    }
  }

  addItemVendor(item: IItemAdd) {
    const itemProduct: IItemProduct = {
      count: 1,
      ...item
    };
    const itemsCheckout: IItemsCheckout = {
      vendor: this.selectedVendor || "",
      items: [itemProduct]
    };
    this.state.addedItems.push(itemsCheckout);
  }

  addItemProduct(item: IItemAdd, index: number) {
    const itemCheckout: IItemsCheckout = this.state.addedItems[index];
    const itemProductIndex = itemCheckout.items.findIndex(
      (itm) => itm.id === item.id
    );
    if (itemProductIndex === -1) {
      const itemProduct: IItemProduct = {
        count: 1,
        ...item
      };
      itemCheckout.items.push(itemProduct);
    } else {
      const itemProduct: IItemProduct = itemCheckout.items[itemProductIndex];
      itemProduct.count += 1;
    }
  }

  onAddItem(item: IItemAdd) {
    const index = this.state.addedItems.findIndex(
      (itm) => itm.vendor === this.selectedVendor
    );
    if (index === -1) {
      this.addItemVendor(item);
    } else {
      this.addItemProduct(item, index);
    }
    this.setState({ addedItems: this.state.addedItems });
  }

  onVendorSelect(vendor: string) {
    this.selectedVendor = vendorData.find((vd) => vd.value === vendor)?.value;
  }

  render() {
    return (
      <div>
        <Select
          data={vendorData}
          label="Vendor"
          onSelect={(vendor) => this.onVendorSelect(vendor)}
          selected={vendorData[0].value}
        />
        <Table
          data={tableData}
          isAddable={true}
          onAddItem={(item) => this.onAddItem(item)}
        />
        <AddRule />
        <Dialog
          buttonText="Check out"
          title="Product Checkout"
          content={
            <Container>
              <List
                items={this.state.addedItems.map((itemCheckout, index) => {
                  return (
                    <ListItemCollapse
                      key={index}
                      header={
                        <ListItemText
                          sx={{ "&": { textTransform: "capitalize" } }}
                        >
                          {itemCheckout.vendor}
                        </ListItemText>
                      }
                      items={
                        <Table
                          data={itemCheckout.items.map((itemCheckout) => ({
                            amount: itemCheckout.count,
                            name: itemCheckout.item.name,
                            price: itemCheckout.item.price
                          }))}
                        />
                      }
                    />
                  );
                })}
              />
            </Container>
          }
        />
      </div>
    );
  }
}

export default ProductCheckoutModule;
