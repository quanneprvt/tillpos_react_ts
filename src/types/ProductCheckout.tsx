import { IItemAdd } from "./Table";

interface IItemProduct extends IItemAdd {
  count: number;
}

interface IItemsCheckout {
  vendor: string;
  items: Array<IItemProduct>;
}

export type { IItemProduct, IItemsCheckout }