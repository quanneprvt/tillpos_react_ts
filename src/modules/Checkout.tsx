import React from "react";
import Dialog from "../components/Dialog";
import ListItemCollapse from "../components/ListItemCollapse";
import Table from "../components/Table";
import List from "../components/List";
import { Container, ListItemText } from "@mui/material";
import { IItemProduct, IItemsCheckout } from "../types/ProductCheckout";
import { IRuleData } from "../types/Rule";
import { operatorSelect, pizzaSelect, productDiscountType } from "../constants";
import { IDiscountResult } from "../types/Checkout";

interface IProps {
  addedItems: Array<IItemsCheckout>;
  rules: Array<IRuleData>;
}

interface IStates {
  price: number;
}

class CheckoutModule extends React.Component<IProps, IStates> {
  prices: Array<number>;
  constructor(props: IProps) {
    super(props);
    this.state = {
      price: 0
    };
    this.prices = [];
  }

  onDialogOpen() {
    this.setState({ price: this.calculatePrice() });
  }

  calculatePrice(): number {
    this.prices = [];
    for (let i = 0; i < this.props.addedItems.length; i++) {
      const vendorRules = this.props.rules.filter(
        (rule) => rule.itemType.vendor === this.props.addedItems[i].vendor
      );
      this.analyzeRule(this.props.addedItems[i].items, vendorRules);
    }
    return this.prices.length > 0 ? this.prices.reduce((a, b) => a + b) : 0;
  }

  analyzeRule(items: IItemProduct[], rules: IRuleData[]) {
    const products: Array<string> = [
      ...Array.from(new Set(items.map((itm) => itm.item.name)))
    ];
    for (let i = 0; i < products.length; i++) {
      const productRules = rules.filter(
        (rule) =>
          pizzaSelect.find((pizza) => pizza.value === rule.itemType.product)
            ?.text === products[i] || rule.itemType.product === "anything"
      );
      const itemToCheck = items.find((itm) => itm.item.name === products[i]);
      const itemIndex = items.findIndex((itm) => itm.item.name === products[i]);
      const discount: IDiscountResult = {
        totalAmount: 0,
        price: 0,
        percent: 0
      };
      for (let j = 0; j < productRules.length; j++) {
        const discountType = productRules[i].discountType.productDiscountType;
        if (discountType === productDiscountType.ALL.value) {
          if (this.checkCondition(itemToCheck, productRules[i])) {
            discount.percent = productRules[i].discountType.percent || 0;
            discount.price = productRules[i].discountType.price || 0;
          }
        } else if (discountType === productDiscountType.EACH.value) {
          if (itemToCheck) {
            discount.totalAmount =
              productRules[i].discountType.amount *
                Math.floor(
                  itemToCheck?.count / productRules[i].itemType.amount
                ) +
              (itemToCheck?.count % productRules[i].itemType.amount);
            if (productRules[i].discountType.price) {
              const normalPrice = itemToCheck.item.price;
              const avgPrice =
                (Math.floor(
                  itemToCheck?.count / productRules[i].itemType.amount
                ) *
                  productRules[i].itemType.amount *
                  productRules[i].discountType.price +
                  (itemToCheck?.count % productRules[i].itemType.amount) *
                    normalPrice) /
                itemToCheck?.count;
              discount.price = avgPrice || 0;
            }
            if (productRules[i].discountType.percent) {
              const normalPercent = 100;
              const avgPercent =
                (Math.floor(
                  itemToCheck?.count / productRules[i].itemType.amount
                ) *
                  productRules[i].itemType.amount *
                  productRules[i].discountType.percent +
                  (itemToCheck?.count % productRules[i].itemType.amount) *
                    normalPercent) /
                itemToCheck?.count;
              discount.percent = avgPercent;
            }
          }
        }
      }
      this.applyRule(items, itemIndex, discount);
    }
  }

  checkCondition(item: IItemProduct | undefined, rule: IRuleData) {
    if (item)
      switch (rule.itemType.operator) {
        case operatorSelect.EQUAL.value:
          return item?.count === rule.itemType.amount;
        case operatorSelect.GREATER_OR_EQUAL_THAN.value:
          return item?.count >= rule.itemType.amount;
        case operatorSelect.GREATER_THAN.value:
          return item?.count > rule.itemType.amount;
        case operatorSelect.LESS_OR_EQUAL_THAN.value:
          return item?.count <= rule.itemType.amount;
        case operatorSelect.LESS_THAN.value:
          return item?.count < rule.itemType.amount;
      }
    else return false;
  }

  applyRule(items: IItemProduct[], itemIndex: number, discount: IDiscountResult) {
    const totalAmount = discount.totalAmount || items[itemIndex].count;
    const price = discount.price || items[itemIndex].item.price;
    const percent = discount.percent || 100;
    console.log(totalAmount, price, percent);
    this.prices.push(totalAmount * price * (percent/100));
  }

  render() {
    return (
      <Dialog
        onDialogOpen={() => this.onDialogOpen()}
        isShowButton={true}
        buttonText="Check out"
        title="Product Checkout"
        content={
          <Container>
            <List
              items={this.props.addedItems.map((itemCheckout, index) => {
                return (
                  <ListItemCollapse
                    key={index}
                    header={
                      <ListItemText
                        sx={{ "&": { textTransform: "capitalize" } }}
                      >
                        {itemCheckout.vendor}
                      </ListItemText>
                    }
                    items={
                      <Table
                        data={itemCheckout.items.map((itemCheckout) => ({
                          amount: itemCheckout.count,
                          name: itemCheckout.item.name,
                          price: itemCheckout.item.price
                        }))}
                      />
                    }
                  />
                );
              })}
            />
            <div style={{marginBottom: "15px"}}>Total Price: {this.state.price}</div>
          </Container>
        }
      />
    );
  }
}

export default CheckoutModule;
