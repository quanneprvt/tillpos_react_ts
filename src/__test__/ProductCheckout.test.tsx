import React from "react";
import ProductCheckout from "../modules/ProductCheckout";
import Checkout from "../modules/Checkout";
import { shallow, ShallowWrapper } from "enzyme";
import { IItemAdd } from "../types/Table";
import { IItemsCheckout } from "../types/ProductCheckout";

const items: Array<IItemAdd> = [
  {
    id: 0,
    item: {
      name: "Small Pizza",
      description: "10'' pizza for one person",
      price: 11.99
    }
  },
  {
    id: 1,
    item: {
      name: "Medium Pizza",
      description: "12'' pizza for two person",
      price: 15.99
    }
  },
  {
    id: 2,
    item: {
      name: "Large Pizza",
      description: "15'' pizza for four person",
      price: 21.99
    }
  }
];

const vendors: Array<string> = ["microsoft", "default", "facebook"];

interface IItemCase {
  id: number;
  count: number;
};
const cases = [
  {
    vendor: "default",
    items: [
      {id: 0, count: 1}, {id: 1, count: 1}, {id: 2, count: 1}
    ]
  },
  {
    vendor: "microsoft",
    items: [
      {id: 0, count: 3}, {id: 2, count: 1}
    ]
  },
  {
    vendor: "amazon",
    items: [
      {id: 1, count: 3}, {id: 2, count: 1}
    ]
  }
];

const results: Array<number> = [49.97, 45.97, 67.96];

describe("ProductCheckout Module", () => {
  let wrapper: ShallowWrapper;
  beforeEach(() => {wrapper = shallow(<ProductCheckout />)});

  it("should render correctly", () => expect(wrapper).toMatchSnapshot());

  it("add 1 basic item", () => {
    const instance = wrapper.instance() as ProductCheckout;
    const testVendor = vendors[0];
    const testItem1 = items[0];
    //
    instance.selectedVendor = testVendor;
    instance.onAddItem(testItem1);
    const add1: IItemsCheckout = {
      vendor: testVendor,
      items: [{
        count: 1,
        ...testItem1
      }]
    };
    expect(instance.state.addedItems).toEqual([add1]);
  });

  it("add some more item", () => {
    const instance = wrapper.instance() as ProductCheckout;
    const testVendor = vendors[0];
    const testItem1 = items[0];
    //
    instance.selectedVendor = testVendor;
    instance.onAddItem(testItem1);
    const add1: IItemsCheckout = {
      vendor: testVendor,
      items: [{
        count: 1,
        ...testItem1
      }]
    };
    expect(instance.state.addedItems).toEqual([add1]);
    const testItem2 = items[1];
    instance.onAddItem(testItem2);
    const add2: IItemsCheckout = {
      vendor: testVendor,
      items: [{count: 1, ...testItem1}, {count: 1, ...testItem2}]
    };
    expect(instance.state.addedItems).toEqual([add2]);
    const testItem3 = items[0];
    instance.onAddItem(testItem3);
    const add3: IItemsCheckout = {
      vendor: testVendor,
      items: [{count: 2, ...testItem1}, {count: 1, ...testItem2}]
    };
    expect(instance.state.addedItems).toEqual([add3]);
  });

  it("price test", () => {
    const instance = wrapper.instance() as ProductCheckout;
    //
    for (let i = 0; i< cases.length; i++) {
      const caseData: IItemCase[] = cases[i].items;
      const testVendor = cases[i].vendor;
      instance.selectedVendor = testVendor;
      for (let j = 0; j< caseData.length; j++) {
        const item = items.find(itm => itm.id === caseData[j].id);
        if (item) {
          for (let k = 0; k< caseData[j].count; k ++)
            instance.onAddItem(item);
        }
      }
      const checkoutWrapper: ShallowWrapper = shallow(<Checkout addedItems={instance.state.addedItems} rules={instance.state.rules}/>);
      const checkoutInstance = checkoutWrapper.instance() as Checkout;
      const price = checkoutInstance.calculatePrice();
      expect(price).toEqual(results[i]);
      instance.setState({ addedItems: [] });
    }
  });
});
