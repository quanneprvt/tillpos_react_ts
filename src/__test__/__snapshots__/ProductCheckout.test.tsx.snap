// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCheckout Module should render correctly 1`] = `ShallowWrapper {}`;
